sql/core/src/main/scala/org/apache/spark/sql/ArrowConverters.scala

```diff
@@ -0,0 +1,432 @@
+    = new NullableFloat4Vector("FloatValue", allocator)
+  override val valueMutator: NullableFloat4Vector#Mutator = valueVector.getMutator
+
+  override def setNull(): Unit = valueMutator.setNull(count)
+  override def setValue(row: InternalRow): Unit
+    = valueMutator.setSafe(count, row.getFloat(ordinal))
+}
+
+private[sql] class DoubleColumnWriter(ordinal: Int, allocator: BaseAllocator)
+    extends PrimitiveColumnWriter(ordinal, allocator) {
+  override val valueVector: NullableFloat8Vector
+    = new NullableFloat8Vector("DoubleValue", allocator)
+  override val valueMutator: NullableFloat8Vector#Mutator = valueVector.getMutator
+
+  override def setNull(): Unit = valueMutator.setNull(count)
+  override def setValue(row: InternalRow): Unit
+    = valueMutator.setSafe(count, row.getDouble(ordinal))
+}
+
+private[sql] class ByteColumnWriter(ordinal: Int, allocator: BaseAllocator)
+    extends PrimitiveColumnWriter(ordinal, allocator) {
+  override val valueVector: NullableUInt1Vector
+    = new NullableUInt1Vector("ByteValue", allocator)
+  override val valueMutator: NullableUInt1Vector#Mutator = valueVector.getMutator
+
+  override def setNull(): Unit = valueMutator.setNull(count)
+  override def setValue(row: InternalRow): Unit
+    = valueMutator.setSafe(count, row.getByte(ordinal))
+}
+
+private[sql] class UTF8StringColumnWriter(ordinal: Int, allocator: BaseAllocator)
+    extends PrimitiveColumnWriter(ordinal, allocator) {
+  override val valueVector: NullableVarBinaryVector
+    = new NullableVarBinaryVector("UTF8StringValue", allocator)
+  override val valueMutator: NullableVarBinaryVector#Mutator = valueVector.getMutator
+
+  override def setNull(): Unit = valueMutator.setNull(count)
+  override def setValue(row: InternalRow): Unit = {
+    val bytes = row.getUTF8String(ordinal).getBytes
+    valueMutator.setSafe(count, bytes, 0, bytes.length)
+  }
+}
+
+private[sql] class BinaryColumnWriter(ordinal: Int, allocator: BaseAllocator)
+    extends PrimitiveColumnWriter(ordinal, allocator) {
+  override val valueVector: NullableVarBinaryVector
+    = new NullableVarBinaryVector("BinaryValue", allocator)
+  override val valueMutator: NullableVarBinaryVector#Mutator = valueVector.getMutator
+
+  override def setNull(): Unit = valueMutator.setNull(count)
+  override def setValue(row: InternalRow): Unit = {
+    val bytes = row.getBinary(ordinal)
+    valueMutator.setSafe(count, bytes, 0, bytes.length)
+  }
+}
+
+private[sql] class DateColumnWriter(ordinal: Int, allocator: BaseAllocator)
+    extends PrimitiveColumnWriter(ordinal, allocator) {
+  override val valueVector: NullableDateVector
+    = new NullableDateVector("DateValue", allocator)
+  override val valueMutator: NullableDateVector#Mutator = valueVector.getMutator
+
+  override def setNull(): Unit = valueMutator.setNull(count)
+  override def setValue(row: InternalRow): Unit = {
+    // TODO: comment on diff btw value representations of date/timestamp
```

Comment:
The Arrow classes are defined more clearly now, so I don't think an explanation will be needed in my next update.